### ISSUE TITLE
Publish fails if fields defining content title are duplicated in Chado

### ISSUE THREAD

**Issue by dsenalik** [tripal/tripal#1924]
### Tripal Version

4.x

### Drupal Version

Probably any

### PostgreSQL Version

Probably any

### PHP Version

Probably any

### Branch
`tv4g1-issue1924-publish-duplicated-titles`

### General Description

We get an error if we try to publish a content type where there is more than one Chado record that will generate the same entity title
For example, I hit this error publishing with two cultivars with the same name.
The default title format for Germplasm Accession is `[germplasm_name]` so there is no unique constraint

### Steps to reproduce

 1. Import Germplasm content type collection
 2. Create two germplasm accessions to be published:
`INSERT into chado.stock (name, uniquename, description, type_id) VALUES ('DupName', 'Unique Name 1', 'DupDescription', (SELECT cvterm_id FROM chado.cvterm WHERE name='accession')), ('DupName', 'Unique Name 2', 'DupDescription', (SELECT cvterm_id FROM chado.cvterm WHERE name='accession'));`
`INSERT 0 2`
```
sitedb=> select * from stock;
 stock_id | dbxref_id | organism_id |  name   |  uniquename   |  description   | type_id | is_obsolete 
----------+-----------+-------------+---------+---------------+----------------+---------+-------------
        1 |           |             | DupName | Unique Name 1 | DupDescription |       3 | f
        2 |           |             | DupName | Unique Name 2 | DupDescription |       3 | f
(2 rows)
```
 3. Publish "Germplasm Accession" content type

### Error messages

```
2024-07-03 08:25:51: Calling: tripal_publish(germplasm, chado_storage, Array)
Step  1 of 6: Find matching records... 
 [notice] Step  1 of 6: Find matching records... 
Step  2 of 6: Generate page titles...
 [notice] Step  2 of 6: Generate page titles...
Step  3 of 6: Find existing published entities...
 [notice] Step  3 of 6: Find existing published entities...
Step  4 of 6: Publishing 2 new entities...
 [notice] Step  4 of 6: Publishing 2 new entities...
Step  5 of 6: Find IDs of entities...
 [notice] Step  5 of 6: Find IDs of entities...
Step  6 of 6: Add field items to published entities...
 [notice] Step  6 of 6: Add field items to published entities...
  Checking for published items for the field: germplasm_dbxref...
 [notice]   Checking for published items for the field: germplasm_dbxref...
  Published 0 items for field: germplasm_dbxref...
 [notice]   Published 0 items for field: germplasm_dbxref...
  Checking for published items for the field: germplasm_dbxref_ann...
 [notice]   Checking for published items for the field: germplasm_dbxref_ann...
  Published 0 items for field: germplasm_dbxref_ann...
 [notice]   Published 0 items for field: germplasm_dbxref_ann...
  Checking for published items for the field: germplasm_description...
 [notice]   Checking for published items for the field: germplasm_description...
ERROR: SQLSTATE[23505]: Unique violation: 7 ERROR:  duplicate key value violates unique constraint "tripal_entity__germplasm_description____pkey"
DETAIL:  Key (entity_id, deleted, delta, langcode)=(13, 0, 0, und) already exists.: 
      INSERT INTO "tripal_entity__germplasm_description"
        (bundle, deleted, entity_id, revision_id, langcode, delta, germplasm_description_record_id) VALUES
(:bundle_1, :deleted_1, :entity_id_1, :revision_id_1, :langcode_1, :delta_1, :germplasm_description_record_id_1),
(:bundle_2, :deleted_2, :entity_id_2, :revision_id_2, :langcode_2, :delta_2, :germplasm_description_record_id_2); Array
(
    [:bundle_1] => germplasm
    [:deleted_1] => 0
    [:entity_id_1] => 13          <***** Both are trying to use 13
    [:revision_id_1] => 1
    [:langcode_1] => und
    [:delta_1] => 0
    [:germplasm_description_record_id_1] => 1
    [:bundle_2] => germplasm
    [:deleted_2] => 0
    [:entity_id_2] => 13          <***** Both are trying to use 13
    [:revision_id_2] => 1
    [:langcode_2] => und
    [:delta_2] => 0
    [:germplasm_description_record_id_2] => 2
)

 [error]  ERROR: SQLSTATE[23505]: Unique violation: 7 ERROR:  duplicate key value violates unique constraint "tripal_entity__germplasm_description____pkey"
DETAIL:  Key (entity_id, deleted, delta, langcode)=(13, 0, 0, und) already exists.: 
      INSERT INTO "tripal_entity__germplasm_description"
        (bundle, deleted, entity_id, revision_id, langcode, delta, germplasm_description_record_id) VALUES
(:bundle_1, :deleted_1, :entity_id_1, :revision_id_1, :langcode_1, :delta_1, :germplasm_description_record_id_1),
(:bundle_2, :deleted_2, :entity_id_2, :revision_id_2, :langcode_2, :delta_2, :germplasm_description_record_id_2); Array
(
    [:bundle_1] => germplasm
    [:deleted_1] => 0
    [:entity_id_1] => 13
    [:revision_id_1] => 1
    [:langcode_1] => und
    [:delta_1] => 0
    [:germplasm_description_record_id_1] => 1
    [:bundle_2] => germplasm
    [:deleted_2] => 0
    [:entity_id_2] => 13
    [:revision_id_2] => 1
    [:langcode_2] => und
    [:delta_2] => 0
    [:germplasm_description_record_id_2] => 2
)
 
[site http://default] [TRIPAL] ERROR: SQLSTATE[23505]: Unique violation: 7 ERROR:  duplicate key value violates unique constraint "tripal_entity__germplasm_description____pkey"DETAIL:  Key (entity_id, deleted, delta, langcode)=(13, 0, 0, und) already exists.:       INSERT INTO "tripal_entity__germplasm_description"        (bundle, deleted, entity_id, revision_id, langcode, delta, germplasm_description_record_id) VALUES(:bundle_1, :deleted_1, :entity_id_1, :revision_id_1, :langcode_1, :delta_1, :germplasm_description_record_id_1),(:bundle_2, :deleted_2, :entity_id_2, :revision_id_2, :langcode_2, :delta_2, :germplasm_description_record_id_2); Array(    [:bundle_1] => germplasm    [:deleted_1] => 0    [:entity_id_1] => 13    [:revision_id_1] => 1    [:langcode_1] => und    [:delta_1] => 0    [:germplasm_description_record_id_1] => 1    [:bundle_2] => germplasm    [:deleted_2] => 0    [:entity_id_2] => 13    [:revision_id_2] => 1    [:langcode_2] => und    [:delta_2] => 0    [:germplasm_description_record_id_2] => 2)
```

**Comment by dsenalik** [comment]
Another example is two publications with the same title. If you have them, they can be found with:
`SELECT title FROM chado.pub GROUP BY title HAVING COUNT(*) > 1;`

I fear this will require a redesign to fix. But now that we have the entity lookup service, we can create a function there to return chado record IDs and corresponding entity IDs, and use this to determine if a record needs to be published, and there will be no need to consult the titles at all 🙏.

**Comment by dsenalik** [comment]
Fixed with merged PR #1991
